File: sdk/java-client/src/test/java/dev/latent/sdk/FakeClient.java
package dev.latent.sdk;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/** Test-only server policy and interface fixture; this is not an SDK transport. */
final class FakeClient implements LatentClient {
    static final class TransportFailure extends RuntimeException {
        TransportFailure(String message) { super(message); }
    }

    static final class ServerRejection extends RuntimeException {
        ServerRejection(String message) { super(message); }
    }

    private static final Models.BudgetConsumption CONSUMPTION =
            new Models.BudgetConsumption(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
    private static final Models.PlatformFailure CANCELLED =
            new Models.PlatformFailure("cancelled", "fixture cancellation", false, List.of());

    static final class Entry {
        final String root;
        final Optional<String> parent;
        final CompletableFuture<Models.InvocationOutcome> response = new CompletableFuture<>();
        Models.ActivationStatus status;
        boolean cancelRequested;

        Entry(String id, Models.InvokeRequest request) {
            root = request.rootActivationId().orElse(id);
            parent = request.parentActivationId();
            status = new Models.ActivationStatus(id, "running", Optional.empty(), Optional.empty(),
                    Optional.empty(), 1, Optional.empty(), Map.of());
        }
    }

    final List<Models.InvokeRequest> requests = new ArrayList<>();
    final Map<String, Entry> entries = new HashMap<>();
    boolean failNextCancel;

    @Override
    public CompletableFuture<Models.InvocationOutcome> invoke(Models.InvokeRequest request) {
        requests.add(request);
        boolean empty = List.of(request.activationId(), request.rootActivationId(), request.parentActivationId())
                .stream().anyMatch(value -> value.isPresent() && value.get().isEmpty());
        if (empty || (request.parentActivationId().isPresent() && request.rootActivationId().isEmpty())) {
            return CompletableFuture.failedFuture(new ServerRejection("invalid-argument"));
        }
        // Assignment belongs to this fake server after observing the unchanged request.
        String id = request.activationId().orElse("server-assigned-" + requests.size());
        if (entries.containsKey(id)) {
            return CompletableFuture.failedFuture(new ServerRejection("already-exists"));
        }
        Entry entry = new Entry(id, request);
        entries.put(id, entry);
        return entry.response;
    }

    @Override
    public CompletableFuture<Models.CancelResponse> cancel(String id, String reason) {
        if (failNextCancel) {
            failNextCancel = false;
            return CompletableFuture.failedFuture(new TransportFailure("cancel transport unavailable"));
        }
        Entry entry = entries.get(id);
        if (entry == null) {
            return CompletableFuture.completedFuture(
                    new Models.CancelResponse(Models.CancelDisposition.NOT_FOUND, Optional.empty()));
        }
        if (entry.status.terminalState().isPresent()) {
            return CompletableFuture.completedFuture(new Models.CancelResponse(
                    Models.CancelDisposition.ALREADY_TERMINAL, entry.status.terminalState()));
        }
        entry.cancelRequested = true;
        return CompletableFuture.completedFuture(
                new Models.CancelResponse(Models.CancelDisposition.ACCEPTED, Optional.empty()));
    }

    @Override
    public CompletableFuture<Models.ActivationStatus> getActivation(String id) {
        Entry entry = entries.get(id);
        return entry == null
                ? CompletableFuture.failedFuture(new ServerRejection("not-found"))
                : CompletableFuture.completedFuture(entry.status);
    }

    void finish(String id, boolean loseResponse) {
        Entry entry = entries.get(id);
        if (entry == null) throw new AssertionError("missing fixture activation");
        String terminal = entry.cancelRequested ? "cancelled" : "completed";
        Models.RetainedInvocationOutcome retained = entry.cancelRequested
                ? new Models.RetainedPlatformFailure(CANCELLED)
                : new Models.ActivationSuccessSummary(Optional.empty(), List.of(), Map.of());
        entry.status = new Models.ActivationStatus(id, entry.cancelRequested ? "running" : "committed", Optional.of(terminal),
                Optional.of(retained), Optional.of(CONSUMPTION), 2, Optional.of(2L), Map.of());
        if (loseResponse) {
            entry.response.completeExceptionally(new TransportFailure("invoke response lost after completion"));
        } else if (entry.cancelRequested) {
            entry.response.complete(new Models.PlatformInvocationFailure(
                    new Models.InvocationReceipt(id, "r", "d", 1, CONSUMPTION), CANCELLED));
        } else {
            entry.response.complete(new Models.InvocationSuccess(new Models.InvokeResponse(
                    id, "r", "d", 1, ByteBuffer.allocate(0), "application/octet-stream",
                    Optional.empty(), List.of(), CONSUMPTION, Map.of())));
        }
    }
}

File: sdk/dotnet/Latent.Sdk.SemanticTests/FakeClient.cs
using Latent.Sdk;

namespace Latent.Sdk.SemanticTests;

// Test-only server policy and interface fixture. This is not an SDK transport.
internal sealed class FakeClient : ILatentClient
{
    internal sealed class TransportFailure(string message) : Exception(message) { }
    internal sealed class ServerRejection(string message) : Exception(message) { }

    private static readonly BudgetConsumption Consumption = new(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
    private static readonly IReadOnlyDictionary<string, string> Metadata = new Dictionary<string, string>();
    private static readonly PlatformFailure Cancelled = new("cancelled", "fixture cancellation", false, []);

    private sealed class Entry(string id, InvokeRequest request)
    {
        public string Root { get; } = request.RootActivationId ?? id;
        public string? Parent { get; } = request.ParentActivationId;
        public TaskCompletionSource<InvocationOutcome> Response { get; } =
            new(TaskCreationOptions.RunContinuationsAsynchronously);
        public ActivationStatus Status { get; set; } = new(id, "running", null, null, null, 1, null, Metadata);
        public bool CancelRequested { get; set; }
    }

    public List<InvokeRequest> Requests { get; } = [];
    public bool FailNextCancel { get; set; }
    private readonly Dictionary<string, Entry> entries = [];

    public ValueTask<InvocationOutcome> InvokeAsync(InvokeRequest request, CancellationToken cancellationToken = default)
    {
        Requests.Add(request);
        if (request.ActivationId == "" || request.RootActivationId == "" || request.ParentActivationId == ""
            || (request.ParentActivationId is not null && request.RootActivationId is null))
        {
            return ValueTask.FromException<InvocationOutcome>(new ServerRejection("invalid-argument"));
        }
        // Assignment belongs to this fake server after observing the unchanged request.
        var id = request.ActivationId ?? $"server-assigned-{Requests.Count}";
        if (entries.ContainsKey(id))
        {
            return ValueTask.FromException<InvocationOutcome>(new ServerRejection("already-exists"));
        }
        var entry = new Entry(id, request);
        entries.Add(id, entry);
        return new ValueTask<InvocationOutcome>(entry.Response.Task);
    }

    public ValueTask<CancelResponse> CancelAsync(string activationId, string reason,
        CancellationToken cancellationToken = default)
    {
        if (FailNextCancel)
        {
            FailNextCancel = false;
            return ValueTask.FromException<CancelResponse>(new TransportFailure("cancel transport unavailable"));
        }
        if (!entries.TryGetValue(activationId, out var entry))
        {
            return ValueTask.FromResult(new CancelResponse(CancelDisposition.NotFound, null));
        }
        if (entry.Status.TerminalState is not null)
        {
            return ValueTask.FromResult(new CancelResponse(CancelDisposition.AlreadyTerminal, entry.Status.TerminalState));
        }
        entry.CancelRequested = true;
        return ValueTask.FromResult(new CancelResponse(CancelDisposition.Accepted, null));
    }

    public ValueTask<ActivationStatus> GetActivationAsync(string activationId,
        CancellationToken cancellationToken = default) =>
        entries.TryGetValue(activationId, out var entry)
            ? ValueTask.FromResult(entry.Status)
            : ValueTask.FromException<ActivationStatus>(new ServerRejection("not-found"));

    public (string Root, string? Parent) Lineage(string id) => (entries[id].Root, entries[id].Parent);

    public void Finish(string id, bool loseResponse = false)
    {
        var entry = entries[id];
        var terminal = entry.CancelRequested ? "cancelled" : "completed";
        RetainedInvocationOutcome retained = entry.CancelRequested
            ? new RetainedInvocationOutcome.PlatformFailure(Cancelled)
            : new RetainedInvocationOutcome.Succeeded(null, [], Metadata);
        entry.Status = new ActivationStatus(id, entry.CancelRequested ? "running" : "committed",
            terminal, retained, Consumption, 2, 2, Metadata);
        if (loseResponse)
        {
            entry.Response.SetException(new TransportFailure("invoke response lost after completion"));
        }
        else if (entry.CancelRequested)
        {
            entry.Response.SetResult(new InvocationOutcome.PlatformFailure(
                new InvocationReceipt(id, "r", "d", 1, Consumption), Cancelled));
        }
        else
        {
            entry.Response.SetResult(new InvocationOutcome.Succeeded(new InvokeResponse(
                id, "r", "d", 1, ReadOnlyMemory<byte>.Empty, "application/octet-stream",
                null, [], Consumption, Metadata)));
        }
    }
}
